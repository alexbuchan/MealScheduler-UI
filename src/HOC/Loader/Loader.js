import React from 'react';
import PropTypes from 'prop-types';

const withLoader = (WrappedComponent) => {
  const propTypes = {
    isLoading: PropTypes.bool,          // Determines whether the loader displays or the wrapped component
    loaderClassName: PropTypes.string,  // Styles class for the loader
    props: PropTypes.object             // Props to be passed to the wrapped component
  };

  const Loader = ({ isLoading, loaderClassName, ...props }) => {
    if (!isLoading) return <WrappedComponent { ...props } />;
    return (
      <div className={ `${loaderClassName}-wrapper` }>
        <h3 className={ loaderClassName } >Loading...</h3>
      </div>
    );
  }

  Loader.propTypes = propTypes;
  return Loader;
}

export default withLoader;