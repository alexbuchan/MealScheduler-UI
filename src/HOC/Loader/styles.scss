.loader-wrapper {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100%;

  .loader {
    border: 10px solid rgba(4, 4, 4, 0.65);
    border-top: 10px solid #ffffff85;
    border-radius: 50%;
    width: 100px;
    height: 100px;
    animation-name: spin;
    animation-duration: 400ms;
    animation-iteration-count: infinite;
    animation-timing-function: linear;
  }

  @keyframes spin {
    from {
        transform:rotate(0deg);
    }
    to {
        transform:rotate(360deg);
    }
  }
}