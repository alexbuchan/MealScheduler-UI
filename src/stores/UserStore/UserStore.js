import Store from '../Store/Store';
import Constants from '../../constants/userConstants';

let userState = {
  auth: false,
  closeFlashMessage: true
};

class UserStore extends Store {
  constructor() {
    super();
  }

  _registerToActions(action) {
    switch(action.actionType) {
      case Constants.REGISTER_USER:
        this.registerUser(action.data);
        break;

      case Constants.LOGIN_USER:
        this.loginUser(action.data);
        break;
      
      case Constants.LOGOUT_USER:
        this.logoutUser(action.data);
        break;

      case Constants.RETRIEVE_USER_DATA_ON_REFRESH:
        this.populateUserDataOnRefresh(action.data);
        break;
    }
  }

  registerUser = data => {
    this.populateUserData(data);
    this.emitChange();
  }

  loginUser = data => {
    this.populateUserData(data);
    this.emitChange();
  }

  populateUserDataOnRefresh = data => {
    this.populateUserData(data);
    this.emitChange();
  }

  logoutUser = (data) => {
    userState.user = null;
    userState.auth = false;
    userState.message = null;
    this.emitChange();
  }

  populateUserData = (data) => {
    userState.auth = data.auth;
    userState.user = data.user;
    userState.message = data.message;
  }

  getUserState() {
    return userState;
  }
}

export default new UserStore();
