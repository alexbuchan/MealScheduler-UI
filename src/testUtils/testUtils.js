import React from 'react';
import { shallow } from 'enzyme';

export const componentSetup = (Component, props = {}, options = {}) => {
  const instance = shallow(<Component { ...props } />, options);
  return instance;
}

export const findByTestAttribute = (instance, attribute) => {
  const wrapper = instance.find(`[data-test="${attribute}"]`);
  return wrapper;
}