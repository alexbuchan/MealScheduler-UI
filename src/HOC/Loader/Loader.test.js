import Loader from './Loader';
import UserStore from '../../stores/UserStore/UserStore';
import { componentSetup } from '../../testUtils/testUtils';

describe('Loader', () => {
  describe('#render', () => {
    describe('when wrapped component is retrieving data', () => {
      let withLoader;
      let instance;
      let props;
    
      beforeEach(() => {
        const mockComponent = jest.fn();
        props = {
          isLoading: true,
          loaderClassName: 'loaderClassName'
        };

        withLoader = Loader(mockComponent);
        instance = componentSetup(withLoader, props);
      });
      
      it('should display the loader', () => {
        expect(instance).toMatchSnapshot();
      });
    });

    describe('when wrapped component is finished retrieving data', () => {
      let withLoader;
      let instance;
      let props;
    
      beforeEach(() => {
        const mockComponent = jest.fn();
        props = {
          isLoading: false,
          loaderClassName: 'loaderClassName'
        };

        withLoader = Loader(mockComponent);
        instance = componentSetup(withLoader, props);
      });

      it('should render the wrapped component', () => {
        const user = { username: 'username', email: 'user@email.com' };
        UserStore.loginUser({ user: user, message: 'User logged in.', auth: true });
        expect(instance).toMatchSnapshot();
      });
    });
  });
});