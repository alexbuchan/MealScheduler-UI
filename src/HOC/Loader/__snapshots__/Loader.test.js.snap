// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader #render when wrapped component is finished retrieving data should render the wrapped component 1`] = `<mockConstructor />`;

exports[`Loader #render when wrapped component is retrieving data should display the loader 1`] = `
<div
  className="loaderClassName-wrapper"
>
  <h3
    className="loaderClassName"
  >
    Loading...
  </h3>
</div>
`;
